include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -564,6 +568,111 @@ __pattern_minmax_element(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __c
     return ::std::make_pair(get<0>(__ret), get<1>(__ret));
 }
 
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+__pattern_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
+                            _Range4&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+{
+    // The algorithm reduces values in [first2, first2 + (last1-first1)) where the
+    // associated keys for the values are equal to the adjacent key.
+    //
+    // Example: keys       = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 } -- [first1, last1)
+    //          values     = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 } -- [first2, first2+n)
+    //
+    //          out_keys   = { 1, 2, 3, 4, 1, 3, 1, 3, 0 } -- result1
+    //          out_values = { 1, 2, 3, 4, 2, 6, 2, 6, 0 } -- result2
+
+    if (__keys.empty())
+        return 0;
+
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+
+    // Round 1: reduce with extra indices added to avoid long segments
+    // TODO: Add a check of whether there are any long key subsequences, and skip a round of copy_if
+    // and reduces if there are none.
```

Comment:
For such a check, you'd probably need to first build a mask for adjacent key equality, then do prefix scan over that mask to determine segment lengths. That does not sounds like an optimization at all :)

Reply:
Thinking further of that, I think we can try something smart. For example, in the first pass for the elements at threshold points we can test the key at the previous threshold point. If it's not the same as ours then the segment has changed in between; if the same, we suspect a long segment, therefore add an intermediate reduction point and set a flag. For the check to be quick, the flag should probably be an atomic variable in USM, set as a side effect in the copy_if predicate.

Reply:
Yes, I agree with that ideas. Now I suggest just adding //TODO about that.

Reply:
I've updated the TODO with the details of Alexey's idea.